README.md
# Responsive jQuery Slider

[![npm](https://img.shields.io/npm/v/codezero-jquery-slider.svg)](https://www.npmjs.com/package/codezero-jquery-slider)
[![npm](https://img.shields.io/npm/dt/codezero-jquery-slider.svg)](https://www.npmjs.com/package/codezero-jquery-slider)
[![npm](https://img.shields.io/npm/l/codezero-jquery-slider.svg)](https://www.npmjs.com/package/codezero-jquery-slider)

[View Demo](https://codezero-be.github.io/jquery-slider/) - [Tinker with CodePen](https://codepen.io/ivanvermeyen/pen/yMWvOR)

Is this the best, most advanced slider out there? No, not at all…
But even with all its features, the slider I was using before didn’t do exactly what I was looking for. So here we are.

## The main goal of this slider is to…
- be able to horizontally traverse through slides,
- allow any HTML content in a slide,
- also be an image slider right out of the box,
- allow slides with fixed and auto widths and heights,
- move 1 slide at a time and center it in the viewport,
- move 70% of the viewport, if the widest slide is less than 30%,
- align slides left, center or right when the viewport is wider than the slides,
- make all of this as customizable as possible…

## Contents
- [Requirements](#requirements)
- [Installation](#installation)
- [Usage](#usage)
  - [Basic Setup](#basic-setup)
  - [Slide Backgrounds](#slide-backgrounds)
  - [Slide Overlays](#slide-overlays)
  - [Image Slides](#image-slides)
    - [Auto Width Image Slides](#auto-width-image-slides)
    - [Full Width Image Slides](#full-width-image-slides)
    - [Covered Width Image Slides](#covered-width-image-slides)
    - [Image Hover Overlay](#image-hover-overlay)
    - [Image Caption](#image-caption)
- [Options](#options)
- [Development](#development)
- [Changelog](#changelog)
- [License](#license)

## Requirements
- [jQuery](https://jquery.com/)
- [jQuery.scrollTo](https://github.com/flesler/jquery.scrollTo)
- [jQuery-Touch-Events](https://github.com/benmajor/jQuery-Touch-Events) (if you want swiping)
- Modern browser (**not** IE 9 or less)

## Installation

### Via Bower
```
bower install codezero-jquery-slider --save
```

### Via NPM
```
npm install codezero-jquery-slider --save
```

### Include JS and CSS
```html
<!-- Important slider styling -->
<link rel="stylesheet" href="css/slider.min.css">

<!-- Required 3rd party scripts -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-scrollTo/2.1.0/jquery.scrollTo.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-touch-events/1.0.8/jquery.mobile-events.js"></script>

<!-- Slider script -->
<script src="js/slider.min.js"></script>

<!-- Kickoff the slider -->
<script>
    $('.slider').slide();
</script>
```

## Usage

### Basic Setup
The following snippet is the base skeleton of a slider.

```html
<div class="slider">

    <a href="#" class="slider-prev">&langle;</a>
    <a href="#" class="slider-next">&rangle;</a>

    <div class="slider-viewport">
        <div class="slider-track">

            <div class="slide">
                <!-- Slide Content -->
            </div>

        </div>
    </div>

</div>
```

By default, each slide will have an **auto width and height**.
You can specify a fixed width and height in CSS yourself,
or [change some variables](https://github.com/codezero-be/jquery-slider/blob/master/src/scss/_variables.scss) if you are compiling SCSS.

If you want every slide to take up 100% width of the viewport, simply add the `slide-full-width` class to the `.slider`.
Or if you want only specific slides to be 100% wide, you can add the `slide-full-width` class to that `.slide`.
The latter is probably less useful...

The slide content can really be anything. Paragraphs, headings, images, ... Style it however you want.

Remember the `.slide` is a `flex` child, so you might want to wrap your content
in another `<div>` for easier styling or positioning...

### Slide Backgrounds
You can [change some variables](https://github.com/codezero-be/jquery-slider/blob/master/src/scss/_variables.scss) to choose default backgrounds for your slides,
write your own styles or use a `data-background` attribute to set a specific background image on a slide.

If you want the background to zoom in when you hover over it, use `data-zoom-background` instead.
You can set your own zoom effect in the [variables file](https://github.com/codezero-be/jquery-slider/blob/master/src/scss/_variables.scss),
as long as it's a valid CSS `transform` value.

```html
<div class="slider slide-full-width">

    <a href="#" class="slider-prev">&langle;</a>
    <a href="#" class="slider-next">&rangle;</a>

    <div class="slider-viewport">
        <div class="slider-track">

            <div class="slide" data-background="path/to/bg-image.jpg">
                <!-- Slide Content -->
            </div>

        </div>
    </div>

</div>
```

### Slide Overlays
To add a color overlay on top of the background image, wrap your slide content in a `<div>` with the `slide-overlay` class.

```html
<div class="slider slide-full-width">

    <a href="#" class="slider-prev">&langle;</a>
    <a href="#" class="slider-next">&rangle;</a>

    <div class="slider-viewport">
        <div class="slider-track">

            <div class="slide" data-background="path/to/bg-image.jpg">
                <div class="slide-overlay">
                    <!-- Slide Content -->
                </div>
            </div>

        </div>
    </div>

</div>
```

### Image Slides
Image slides have specific markup and default styling out of the box.
The only thing you might want to hook up yourself is any lightbox you prefer.

#### Auto Width Image Slides
To restrain the image size, you might want to specify a `max-height` for the `<img>`.
If you set the `$slide-height` [variable in SCSS](https://github.com/codezero-be/jquery-slider/blob/master/src/scss/_variables.scss),
this is done automatically.

```html
<div class="slider">

    <a href="#" class="slider-prev">&langle;</a>
    <a href="#" class="slider-next">&rangle;</a>

    <div class="slider-viewport">
        <div class="slider-track">

            <div class="slide">
                <a href="path/to/original-image.jpg" class="slide-image">
                    <img src="path/to/image-thumb.jpg">
                </a>
            </div>

        </div>
    </div>

</div>
```

> Note that the `.slide-image` doesn't have to be a link.
> It can also just be a `<div>`.

#### Full Width Image Slides
Again, you can make each slide full width by adding the `slide-full-width` class.
The clickable `.slide-image` element will be centered inside the slide.

```html
<div class="slider slide-full-width">

    <a href="#" class="slider-prev">&langle;</a>
    <a href="#" class="slider-next">&rangle;</a>

    <div class="slider-viewport">
        <div class="slider-track">

            <div class="slide">
                <a href="path/to/original-image.jpg" class="slide-image">
                    <img src="path/to/image-thumb.jpg">
                </a>
            </div>

        </div>
    </div>

</div>
```

#### Covered Width Image Slides
To make the `.slide-image` element also full width, replace the `slide-full-width` class with `slide-cover-width`. 
Now only the image is still centered inside the `.slide-image` element, which makes sense since we don't want to stretch it out.

However, to set the image as a background, covering the entire slide,
simply add the `slide-image-background` class to the `<img>`.
Note that this will probably crop the image to fit the slides dimensions.

```html
<div class="slider slide-cover-width">

    <a href="#" class="slider-prev">&langle;</a>
    <a href="#" class="slider-next">&rangle;</a>

    <div class="slider-viewport">
        <div class="slider-track">

            <div class="slide">
                <a href="path/to/original-image.jpg" class="slide-image">
                    <img src="path/to/image-thumb.jpg" class="slide-image-background">
                </a>
            </div>

        </div>
    </div>

</div>
```

#### Image Hover Overlay
Want to show a zoom icon when you hover over the image? Just add it...

```html
<div class="slide">
    <a href="path/to/original-image.jpg" class="slide-image">
        <!-- Note: this icon requires Font Awesome... -->
        <!-- Use any library you like... -->
        <i class="fa fa-search-plus slide-image-hoverlay"></i>
        <img src="path/to/image-thumb.jpg">
    </a>
</div>
```
 
 Or just show some text...
 
```html
<div class="slide">
    <a href="path/to/original-image.jpg" class="slide-image">
        <span class="slide-image-hoverlay">VIEW</span>
        <img src="path/to/image-thumb.jpg">
    </a>
</div>
```

#### Image Caption
Same as the hoverlay, if you would like to display an image caption, just add the element...

```html
<div class="slide">
    <a href="path/to/original-image.jpg" class="slide-image">
        <img src="path/to/image-thumb.jpg">
        <span class="slide-image-caption">
            Beautiful example image!
        </span>
    </a>
</div>
```

## Options
If you are using SASS/SCSS, there are a bunch of variables you can tweak to change the behavior of the slider.
I have added comments, so I hope it will be clear what each option will do.
Just take a look at the [variables file](https://github.com/codezero-be/jquery-slider/blob/master/src/scss/_variables.scss).

The javascript side also has some options. These are also commented so I would again suggest to
take a look at the `defaults` object in the [javascript source file](https://github.com/codezero-be/jquery-slider/blob/master/src/js/slider.js).

Overriding the defaults would take the form of:

```javascript
$('.slider').slide({
    slideSpeed: 300,
    enableSwipe: false
});
```

## Development
I am using a tool called [Laravel Mix](https://github.com/JeffreyWay/laravel-mix) to compile javascript and SCSS.
Make sure you have the latest version of [NodeJS](https://nodejs.org/en/) installed and then run `npm install`.

- To compile run `npm run dev`.
- To compile and watch for changes run `npm run watch`.
- To compile for production (minify/uglify) run `npm run production`.

## Changelog

View the [changelog](https://github.com/codezero-be/jquery-slider/blob/master/CHANGELOG.md).

## License
The MIT License (MIT). Please see [License File](https://github.com/codezero-be/jquery-slider/blob/master/LICENSE.md) for more information.

---
[![Analytics](https://ga-beacon.appspot.com/UA-58876018-1/codezero-be/jquery-slider)](https://github.com/igrigorik/ga-beacon)


CHANGELOG.md
# Changelog

All Notable changes will be documented in this file.

## 1.0.2 (2017-04-21)

- Minify compiled JS and CSS files

## 1.0.1 (2017-04-20)

- Add slide overlay info to readme
- Update demo page

## 1.0.0 (2017-04-20)

- Initial release


src/js/slider.js
(function ($, window, document, undefined) {

    var pluginName = 'slide',
        defaults = {
            // General settings...
            slideSpeed: 500,
            enableSwipe: true,

            // If you change class / data attribute names,
            // you will need to change related CSS files
            viewport: '.slider-viewport',
            track: '.slider-track',
            slide: '.slide',
            // Arrows
            prevArrow: '.slider-prev',
            nextArrow: '.slider-next',
            // Slider state
            atLastSlide: '.slider-end',
            atFirstSlide: '.slider-start',
            noSlide: '.no-slide',
            // Slide image classes
            imageContainerClass: '.slide-image',
            imageAsBackgroundClass: '.slide-image-background',
            imageAsBackgroundWrapperClass: '.slide-image-background-wrapper',
            // Slide background classes / data attributes
            backgroundClass: '.slide-data-background',
            backgroundZoomClass: '.slide-data-zoom-background',
            backgroundDataAttr: 'background',
            backgroundZoomDataAttr: 'zoom-background',

            // Check if we should enable single slide mode..
            // Return true to scroll only one slide or false to slide the default distance.
            // You can also set this to a boolean instead of a function.
            // By default, if any slide is wider than 30% of the viewport, single slide mode is enabled.
            isInSingleSlideMode: function ($slider) {
                var isInSingleSlideMode = false,
                    viewportWidth = $slider.find(this.viewport).width();

                $slider.find(this.slide).each(function () {
                    isInSingleSlideMode = $(this).outerWidth() / viewportWidth > .3;
                    return ! isInSingleSlideMode;
                });

                return isInSingleSlideMode;
            },

            // Slide distance used if "isInSingleSlideMode" is true.
            // Return any value supported by the jquery.scrollTo plugin:
            // https://github.com/flesler/jquery.scrollTo
            // By default this will slide 70% of the viewport.
            defaultSlideDistance: function ($slider, $viewport, $track, isNext) {
                return (isNext ? '+=' : '-=') + ($viewport.width() * .7) + 'px';
            },

            // Before callbacks...
            // Return false to cancel slide.
            onBeforeSlideNext: function ($slider) { },
            onBeforeSlidePrev: function ($slider) { },

            // After callbacks...
            onAfterSlideNext: function ($slider) { },
            onAfterSlidePrev: function ($slider) { }
        };

    function Plugin(element, options) {
        // Merge options...
        this.options = $.extend( {}, defaults, options);

        // Cache elements...
        this.$slider = $(element);
        this.$viewport = this.$slider.find(this.options.viewport);
        this.$track = this.$slider.find(this.options.track);
        this.$slides = this.$slider.find(this.options.slide);

        // Calculated values...
        this.viewportWidth = 0;
        this.slidesTotalWidth = 0;
        this.singleSlideIsWiderThanViewport = false;
        this.slidesFitInViewport = false;
        this.isInSingleSlideMode = false;
        this.noSlideClass = (this.options.noSlide).substr(1);
        this.onResize = null;
        this.isSliding = false;

        // Kickoff...
        this.init();
    }

    Plugin.prototype = {

        init: function ()  {
            this.swapSlideCoverImages();
            this.insertDataBackgrounds();
            this.registerEvents();
            this.evaluateSlider();

            // Do a recheck after 1 second
            // in case images load slowly...
            setTimeout(function () {
                this.evaluateSlider();
            }.bind(this), 1000);
        },

        swapSlideCoverImages: function () {
            this.$slider.find('img' + this.options.imageAsBackgroundClass).each(function (index, image) {
                var $image = $(image),
                    $container = $image.closest(this.options.imageContainerClass),
                    imageUrl = $image.prop('src');

                if (imageUrl) {
                    $container
                        .css('backgroundImage', 'url(' + imageUrl + ')')
                        .addClass((this.options.imageAsBackgroundWrapperClass).substr(1));
                }
            }.bind(this));
        },

        insertDataBackgrounds: function () {
            this.$slider.find(this.options.slide).each(function (index, slide) {
                var $slide = $(slide),
                    $background,
                    backgroundUrl = $slide.data(this.options.backgroundDataAttr) || $slide.data(this.options.backgroundZoomDataAttr),
                    shouldZoom = !! $slide.data(this.options.backgroundZoomDataAttr);

                if (backgroundUrl) {
                    $background = $('<div/>')
                        .addClass((this.options.backgroundClass).substr(1))
                        .css('backgroundImage', 'url(' + backgroundUrl + ')');

                    if (shouldZoom) {
                        $background.addClass((this.options.backgroundZoomClass).substr(1));
                    }

                    $slide.prepend($background);
                }
            }.bind(this));
        },

        registerEvents: function () {
            // Next arrow click...
            this.$slider.on('click', this.options.nextArrow, function (e) {
                e.preventDefault();
                this.slideTo(this.$slides, true);
            }.bind(this));

            // Prev arrow click...
            this.$slider.on('click', this.options.prevArrow, function (e) {
                e.preventDefault();
                this.slideTo($(this.$slides.get().reverse()), false);
            }.bind(this));

            if (this.options.enableSwipe) {
                // Swipe left...
                this.$slider.on('swiperight', function () {
                    this.slideTo($(this.$slides.get().reverse()), false);
                }.bind(this));

                // Swipe right...
                this.$slider.on('swipeleft', function () {
                    this.slideTo(this.$slides, true);
                }.bind(this));

                // No dragging when "swiping" with the mouse...
                this.$slider.on('dragstart', 'a, img', function (e) {
                    e.preventDefault();
                });

                // Don't follow links when swiping (IE 11 & Edge)...
                this.$slider.on('click', 'a', function (e) {
                    if (this.isSliding) {
                        e.preventDefault();
                    }
                }.bind(this));
            }

            // Window resize event...
            $(window).on('resize', function () {
                clearTimeout(this.onResize);
                this.onResize = setTimeout(function () {
                    this.evaluateSlider();
                    this.onResize = null;
                }.bind(this), 900);
            }.bind(this));
        },

        // Triggered on init
        // and on window resize.
        evaluateSlider: function () {
            this.updateSliderInfo();
            this.updateSlider();
            this.updateArrows();
        },

        updateSliderInfo: function () {
            this.viewportWidth = this.getViewportWidth();
            this.slidesTotalWidth = this.getSlidesWidth();
            this.singleSlideIsWiderThanViewport = this.isSingleSlideWiderThanViewport();
            this.slidesFitInViewport = this.checkSlidesFitInViewport();
            this.isInSingleSlideMode = this.options.isInSingleSlideMode instanceof Function
                ? this.options.isInSingleSlideMode(this.$slider)
                : this.options.isInSingleSlideMode;
        },

        updateSlider: function () {
            if (this.slidesFitInViewport || this.singleSlideIsWiderThanViewport) {
                this.$slider.addClass(this.noSlideClass);
            } else {
                this.$slider.removeClass(this.noSlideClass);
            }

            if (this.singleSlideIsWiderThanViewport) {
                this.slideTo(this.$slides, true);
            }
        },

        updateArrows: function () {
            var atLastSlide = (this.options.atLastSlide).substr(1),
                atFirstSlide = (this.options.atFirstSlide).substr(1);

            if (this.isAtLastSlide()) {
                this.$slider.addClass(atLastSlide);
            } else {
                this.$slider.removeClass(atLastSlide);
            }

            if (this.isAFirstSlide()) {
                this.$slider.addClass(atFirstSlide);
            } else {
                this.$slider.removeClass(atFirstSlide);
            }
        },

        slideTo: function ($slides, isNext) {
            if (this.isSliding || this.runBeforeCallback(isNext) === false) {
                return false;
            }

            this.isSliding = true;

            this.$viewport.scrollTo(this.getSlideToPosition($slides, isNext), this.options.slideSpeed, {
                onAfter: function () {
                    this.updateArrows();
                    this.runAfterCallback(isNext);
                    this.isSliding = false;
                }.bind(this)
            });
        },

        getSlideToPosition: function ($slides, isNext) {
            if ( ! this.isInSingleSlideMode) {
                return this.options.defaultSlideDistance(this.$slider, this.$viewport, this.$track, isNext);
            }

            var trackOffset = this.getTrackOffset(),
                halfViewportWidth = this.viewportWidth / 2,
                slideToOffset = 0,
                isPrev = ! isNext;

            $slides.each(function (index, slide) {
                var $slide = $(slide),
                    slideWidth = $slide.outerWidth(),
                    leftOffset = $slide.position().left + parseInt($slide.css("marginLeft")),
                    slideCenterPosition = leftOffset + (slideWidth / 2) - trackOffset,
                    slideCenterIsOverHalfWay = slideCenterPosition - 2 > halfViewportWidth,
                    slideCenterIsBeforeHalfWay = slideCenterPosition + 2 < halfViewportWidth;

                slideToOffset = leftOffset + ((slideWidth - this.viewportWidth) / 2); //=> Center slide

                if ( (isNext && slideCenterIsOverHalfWay) ||
                     (isPrev && slideCenterIsBeforeHalfWay) ) {
                    return false;
                }
            }.bind(this));

            return slideToOffset;
        },

        getTrackOffset: function () {
            return Math.abs(this.$track.position().left);
        },

        getViewportWidth: function () {
            return parseFloat(this.$viewport.width());
        },

        getSlidesWidth: function () {
            var width = 0;

            this.$slides.each(function () {
                width += parseFloat($(this).outerWidth(true));
            });

            return width;
        },

        checkSlidesFitInViewport: function () {
            return this.viewportWidth > this.slidesTotalWidth;
        },

        isSingleSlideWiderThanViewport: function () {
            return this.$slides.length <= 1 && this.slidesTotalWidth >= this.viewportWidth;
        },

        isAFirstSlide: function () {
            return this.getTrackOffset() - 1 <= this.getSlideOverflow(this.$slides.first());
        },

        isAtLastSlide: function () {
            var trackRemaining = this.slidesTotalWidth - this.getTrackOffset() - 1,
                slideOverflow = this.getSlideOverflow(this.$slides.last());

            return this.viewportWidth >= trackRemaining - slideOverflow;
        },

        getSlideOverflow: function ($slide) {
            if ($slide.outerWidth() <= this.viewportWidth) {
                return 0;
            }

            return ($slide.outerWidth() - this.viewportWidth) / 2;
        },

        runBeforeCallback: function (isNext) {
            var beforeCallback = isNext
                    ? this.options.onBeforeSlideNext
                    : this.options.onBeforeSlidePrev;

            if (beforeCallback instanceof Function) {
                return beforeCallback(this.$slider);
            }

            return true;
        },

        runAfterCallback: function (isNext) {
            var afterCallback = isNext
                    ? this.options.onAfterSlideNext
                    : this.options.onAfterSlidePrev;

            if (afterCallback instanceof Function) {
                afterCallback(this.$slider);
            }
        }
    };

    $.fn[pluginName] = function (options) {
        return this.each(function () {
            if ( ! $.data(this, "plugin_" + pluginName)) {
                $.data(this, "plugin_" + pluginName,
                new Plugin(this, options));
            }
        });
    };

})(jQuery, window, document);
